### PR TITLE
cli: set default verbosity level with environment variable

### PR DESCRIPTION
**Pull request by mr-cal**
- [X] Have you followed the [guidelines for contributing](https://github.com/snapcore/snapcraft/blob/master/CONTRIBUTING.md)?
- [X] Have you signed the [CLA](http://www.ubuntu.com/legal/contributors/)?
- [x] Have you successfully run `make lint`?
- [x] Have you successfully run `pytest tests/unit`?

-----

The default verbosity level can be set with the environment variable `SNAPCRAFT_VERBOSITY_LEVEL`, where valid values are `quiet`, `brief`, `verbose`, `debug`, and `trace`.

The implementation I chose for this PR isn't correct.  Running `SNAPCRAFT_VERBOSITY_LEVEL=trace snapcraft pull --verbosity=quiet` produces the output
```
2022-10-24 15:36:28.919 Starting Snapcraft 7.1.1.post1+git6934c0ba
2022-10-24 15:36:28.919 Logging execution to '/home/developer/.cache/snapcraft/log/snapcraft-20221024-153628.919185.log'
```

Running `SNAPCRAFT_VERBOSITY_LEVEL=trace snapcraft pull --verbosity=trace` outputs the greeting message twice:
```
2022-10-24 15:36:44.269 Starting Snapcraft 7.1.1.post1+git6934c0ba
2022-10-24 15:36:44.269 Logging execution to '/home/developer/.cache/snapcraft/log/snapcraft-20221024-153644.268436.log'
2022-10-24 15:36:44.269 Starting Snapcraft 7.1.1.post1+git6934c0ba
2022-10-24 15:36:44.269 Logging execution to '/home/developer/.cache/snapcraft/log/snapcraft-20221024-153644.268436.log'
2022-10-24 15:36:44.269 Raw pre-parsed sysargs: args={'help': False, 'verbose': False, 'quiet': False, 'verbosity': 'trace', 'version': False, 'trace': False} filtered=['pull']
2022-10-24 15:36:44.269 General parsed sysargs: command='pull' args=[]
... (logs continue in 'trace' mode)
```

I'd like to talk to @facundobatista this week about a better solution.  I'm thinking about adding an optional parameter to `craft_cli.dispatcher.pre_parse_args()`.

Another option would be allowing snapcraft to call `pre_parse_args()` (or another subset of this method) to parse the verbosity level before initializing the emitter.

Update: Facundo and I talked about a solution and raised a non-blocking craft-cli issue: https://github.com/canonical/craft-cli/issues/128

(CRAFT-1420)

### CODE REVIEW

File: snapcraft/cli.py [hunk 1]
```diff
@@ -148,6 +148,23 @@ def get_verbosity() -> EmitterMode:
...
+    #      `set_mode()`, which logs the greeting twice.
+
+    # if defined, use environmental variable SNAPCRAFT_VERBOSITY_LEVEL
+    verbosity_env = os.getenv("SNAPCRAFT_VERBOSITY_LEVEL")
+    if verbosity_env:
+        try:
+            verbosity = EmitterMode[verbosity_env.strip().upper()]
+        except KeyError:
+            raise ArgumentParsingError(
+                f"cannot parse verbosity level {verbosity_env!r} from environmental "
```
Comment (sergiusens) [inline snapcraft/cli.py:163]:
```suggestion
                f"cannot parse verbosity level {verbosity_env!r} from environment "
```

File: snapcraft/cli.py [hunk 2]
```diff
@@ -148,6 +148,23 @@ def get_verbosity() -> EmitterMode:
...
+                "variable SNAPCRAFT_VERBOSITY_LEVEL (valid values are 'quiet', "
+                "'brief', 'verbose', 'debug' and 'trace')"
```
Comment (sergiusens) [inline snapcraft/cli.py:165]: you can look into doing a humanize join of the existing enum to get the valid values

**Comment by mr-cal** [comment]
Spread failure unrelated (https://github.com/snapcore/snapcraft/pull/3958)

File: snapcraft/utils.py [hunk 3]
```diff
@@ -290,19 +290,27 @@ def prompt(prompt_text: str, *, hide: bool = False) -> str:
...
+    item_format: str = "{!r}",
+    sort: bool = True,
 ) -> str:
     """Format a list into a human-readable string.
 
     :param items: list to humanize.
     :param conjunction: the conjunction used to join the final element to
                         the rest of the list (e.g. 'and').
     :param item_format: format string to use per item.
+    :param sort: if true, sort the list
```
Comment (tigarmo) [inline snapcraft/utils.py:304]:
```suggestion
    :param sort: if true, sort the list.
```

File: snapcraft/utils.py [hunk 4]
```diff
@@ -290,19 +290,27 @@ def prompt(prompt_text: str, *, hide: bool = False) -> str:
 
 
 def humanize_list(
-    items: Iterable[str], conjunction: str, item_format: str = "{!r}"
+    items: Iterable[str],
+    conjunction: str,
+    item_format: str = "{!r}",
+    sort: bool = True,
```
Comment (tigarmo) [inline snapcraft/utils.py:296]: I think we should add a quick test for this new parameter

**Review by tigarmo — CHANGES_REQUESTED** [review]

File: snapcraft/utils.py (hunk 4 above)
Comment (mr-cal) [reply snapcraft/utils.py:296]: Good catch!

**Review by tigarmo — APPROVED** [review]

**Pull request merged**